src/main/java/frc/robot/subsystems/targetvision/TargetVision.kt

```diff
@@ -1,61 +1,203 @@
+import kotlin.math.pow
+import edu.wpi.first.math.Vector
+
+
+object TargetVision:  Subsystem {
 
-object TargetVision : Subsystem {
     private val io: TargetVisionIO = Limelight
-    private val inputs: TargetVisionIO.Inputs = TargetVisionIO.Inputs()
+    private val inputs: TargetVisionIO.TargetVisionIOInputs = TargetVisionIO.TargetVisionIOInputs()
     private const val SAMPLE_NUM = 8
 
+
+    private fun getDistance(target: LimelightTarget_Detector): Double{
+        val angleToBucket = Units.degreesToRadians(CAMERA_PITCH + target.ty)
+        return (BUCKET_HEIGHT - LIMELIGHT_HEIGHT) / tan(angleToBucket)
+    }
+
+    //ascending
+    private val targetsbyDistance: List<Pair<LimelightTarget_Detector, Double>>
+        get() {
+            return inputs.targets.map {target: LimelightTarget_Detector -> Pair(target, getDistance(target))}.sortedBy {it.second}
+        }
+    val hasTargets: Boolean
+        get(){
+            return inputs.hasTargets
+        }
+    val curTime: Double
+        get() {return inputs.lastTimeStampMS}
+
+
+
     override fun periodic() {
-        io.updateInputs(inputs)
-        Logger.getInstance().processInputs("Target Vision", inputs)
+        Logger.getInstance().processInputs("Vision", inputs)
 
-        if (inputs.targets.isNotEmpty()) {
-            addMeasurement(takeMeasurement(targetsByDistance[0].first))
+        for (target in targetsbyDistance) {
+            if(target.first.classID == (OpposingAlliance.id)){
+                
+                addMeasurement(takeMeasurement(target.first))
+            }
         }
+
+        io.updateInputs(inputs)
     }
 
-    // in ascending order
-    private val targetsByDistance: List<Pair<LimelightTarget_Retro, Double>>
-        get() {
-            return inputs.targets.map { target: LimelightTarget_Retro -> Pair(target, distanceToTarget(target)) }
-                .sortedBy { it.second }
+   data class Measurement(val timestamp: Double, val pose: LimelightTarget_Detector)
+
+    data class Sample(val timestamp : Double, val pose: Translation2d)
+
+    private fun takeMeasurement(target: LimelightTarget_Detector): Measurement{
+
+       //val targetTranslation = Translation2d(getDistance(target), target.tx)
+
+       return Measurement(inputs.lastTimeStampMS, target)
+
+   }
+
+
+
+    val primaryTargetSamples: List<Sample>
+        get(){
+            return robotPoses[0].map { pair -> Sample(pair.timestamp, Translation2d(getDistance(pair.pose), pair.pose.tx))}
+        }
+
+
+    val primaryTarget: LimelightTarget_Detector
+        get(){
+            return robotPoses[0].last().pose
         }
 
-    private var measurements: MutableList<Measurement> = mutableListOf()
 
-    data class Measurement(val timestamp: Double, val translation: Translation2d)
+    var robotPoses: MutableList<MutableList<Measurement>> = mutableListOf()
+
+
+    private fun groupConfidence(last: LimelightTarget_Detector, measurement: LimelightTarget_Detector): Double
```

Comment:
Should probably be comments explaning the algorithm, although that can wait for when we're sure we're using this algo.